### ISSUE TITLE
Document usage

### ISSUE THREAD

**Issue by jasonbahl** [wp-graphql/wp-graphql-jwt-authentication#6]
Need to document the flow and how to use JWT tokens. 

Basic flow is: 

## Initial Request to Authenticate
Send GraphQL Mutation to login with username and password, asking for an `authToken` and `refreshToken` in response. Store tokens for use in future requests. The `authToken` and `refreshToken`

## Authenticated Requests
* Before making an Authenticated request, check to see if the `authToken` is expired.
* If not expired, Use the `authToken` in the request headers: 
  * `"Authorization": "Bearer $authTokenGoesHere"`
* If the token is valid, the user will be set as the user ID in the token and the request will be processed with that user set, resolving public data and private data that user has capabilities to resolve.
  * Additionally, in the response headers will be a `refreshToken`. The refresh token can be used later to get an updated `authToken`. 
* If the token is expired or in any other way invalid, the response will resolve with no user set, so public data will be returned, but private data will not, and the status will be set to 403 indicating that the Auth was invalid. The client should use the 403 response as indication that a new `authToken` is needed. 

## Generating new Auth Tokens
* A new `authToken` can be retrieved by:
  * Sending a valid refresh token in the headers:
     * `"Authorization-Refresh": "$refreshTokenGoesHere"`
     * A request that contains a valid refresh token in the headers will provide a new `authToken` in the response headers as:  `X-JWT-Authentication`. This new `authToken` can be used for subsequent requests until it's expired. 
  * Or by sending another login mutation).

### What makes a valid token?
* In short, an authToken is valid once it's issued until it expires (usually a short expiration date) and a refresh token is valid until it's revoked or expires (usually long expiration date).
* An `authToken` is essentially considered valid if the signature can be verified and it's not expired.
* A `refreshToken` is tied to a `user secret` that can be revoked, allowing for all previously issued `refreshToken` to be invalidated if the user-secret is refreshed or revoked. The user-secret is automatically refreshed anytime a user resets their password, and can also be refreshed or revoked manually by the user or an admin with proper capabilities to edit users.

**Comment by dre1080** [comment]
@jasonbahl Hi, can you give a better example of this? I'm very confused on how to use this.

**Comment by jasonbahl** [comment]
@dre1080 I'd suggest checking out the tests to see how this is meant to be used. 

I do have a new project I'm working on to better showcase features like authentication, etc. . .you can follow that project here, but it's super new, and nothing there quite yet. . .once I have some stuff ironed out locally I'll push it up: https://github.com/wp-graphql/dashboard.wpgraphql.com

**Comment by jasonbahl** [comment]
@dre1080 specifically these functional tests: https://github.com/wp-graphql/wp-graphql-jwt-authentication/tree/master/tests/functional

**Comment by stug111** [comment]
@jasonbahl Hi, trying to create a Gatby + Apollo + Wocommerce website. There is a problem with authentication. I login and after a few minutes the server gives 403 error. How to implement a token update?

Code can be posted [here](https://github.com/stug111/GatsbyApolloWoocommerce).

**Comment by dre1080** [comment]
@jasonbahl thanks, but is there any real world example application showing integrating these tests with something like Apollo Client to show the full login/logout flow and "remember me" logins? Showing Refresh/Revoke Token usage etc.
